### PR TITLE
Feature/issue 21/extends issue 22 user registration login

### PR DESCRIPTION
**Pull request by TekuriSaiAkhil**
# [Issue 21](https://github.com/gcivil-nyu-org/wed-fall24-team5/issues/21) + [Issue 22](https://github.com/gcivil-nyu-org/wed-fall24-team5/issues/22)

## Description

- Implemented User Registration
- Implemented User Login
- Added Login with Google

## Change Type (delete non-relevant options)
- 💡 New feature (non-breaking change which adds functionality)

## Dependencies

- To try this out locally:
1. Activate your virtual env and install all dependencies in requirements.txt
2. add a .env file to src/ 
3. cd src
4. python manage.py makemigrations
5. python manage.py migrate
6. [optional] python manage.py createsuperuser  
7. python manage.py runserver

### CODE REVIEW

File: src/django_management/settings.py
Comment (maxgrove42) [inline src/django_management/settings.py:31]: Could you add a comment why this is set to `3`?

File: src/django_management/settings.py
Comment (maxgrove42) [inline src/django_management/settings.py:42]: for these - could you add a comment to separate out what these are for ? 

```python
     # google login providers
     'django.contrib.sites',
     'allauth',
     'allauth.account',
     'allauth.socialaccount',
     'allauth.socialaccount.providers.google',
     #other apps
     'accounts',
```

File: src/django_management/settings.py
Comment (maxgrove42) [inline src/django_management/settings.py:57]: same here - comment for what this is for would be great ! sorry to be nitpick on that, just so we can understand why this was added.

**Review by maxgrove42 — CHANGES_REQUESTED** [review]
We're failing the linting checks, which are used to ensure clean code styles. If you go to the Pull Request, then the `Checks` tab, you'll see the two checks. One for `Lint` and one for `Branch Name Check`. 

In order for the pull request to be merge-able, it needs to pass all the linting checks. You can also run the linting locally using the instructions in the `README` file.

File: src/django_management/urls.py
Comment (maxgrove42) [inline src/django_management/urls.py:23]: it looks like this is causing the error - when we navigate to `localhost/`, it auto redirects to `allauth.urls` which doesn't exist as a module in the code.

I also imagine an error would occur since we have to paths to the same path (both to `''` or the `localhost/` site.

Comment (TekuriSaiAkhil) [reply src/django_management/urls.py:23]: Please go migrations before starting the server.

Comment (TekuriSaiAkhil) [reply src/django_management/urls.py:23]: Please check the PR dependencies in the description. I will go over this in next meeting.

Comment (maxgrove42) [reply src/django_management/urls.py:23]: Same error on `python src/manage.py migrate` and `python src/manage.py makemigrations`

ModuleNotFoundError: No module named 'allauth'

**Comment by maxgrove42** [comment]
Also ! I'll be excited to hear about the google admin setup. We should try to figure out how to get that in Github Actions and the Github Secret Keys / Environment variables.

File: src/django_management/urls.py
Comment (TekuriSaiAkhil) [reply src/django_management/urls.py:23]: I've updated requirements.txt. Please run `pip install -r requirements.txt` in your virtualenv.

Comment (maxgrove42) [reply src/django_management/urls.py:23]: nice that worked ! great work!! 

let's hold off on merging this until we can get the google auth setup on our locals and potentially get that into a GitHub environment variable.

File: src/accounts/forms.py
Comment (maxgrove42) [inline src/accounts/forms.py:71]: are username and email ever going to be different?

and is password1 different than password2 ? is this just in the instance a user has an email login and a google account login ?

Comment (maxgrove42) [inline src/accounts/forms.py:71]: do we ever authenticate on signup that `password1 == password2` ? I feel that this should be done on the front-end to prevent having to send an invalid form to the back-end.

File: src/django_management/settings.py
Comment (TekuriSaiAkhil) [reply src/django_management/settings.py]: Updated the code to automate the Google login setup. Reopen if required.

File: src/django_management/settings.py
Comment (TekuriSaiAkhil) [reply src/django_management/settings.py:45]: These are apps required for social auth. This is like a template that we get in Django's initial setup.

File: src/django_management/settings.py
Comment (TekuriSaiAkhil) [reply src/django_management/settings.py:60]: same herer - These are apps required for social auth. This is like a template that we get in Django's initial setup.

File: src/accounts/forms.py
Comment (TekuriSaiAkhil) [reply src/accounts/forms.py:71]: This is how Django does it on the back end.

File: src/django_management/settings.py
Comment (maxgrove42) [reply src/django_management/settings.py:45]: I think it'd be nice to include a comment saying that these are for social auth just to help keep things organized and explain what the different `installed_apps` are for

File: src/django_management/settings.py
Comment (maxgrove42) [reply src/django_management/settings.py:60]: Same here - I think it'd be nice to include a comment saying that these are for social auth just to help keep things organized and explain what the different `middlewares` are for.

Also if you don't mind keep the comments unresolved until the original author reviews them, I don't see your responses once you resolve

File: src/django_management/urls.py
Comment (maxgrove42) [inline src/django_management/urls.py:24]: nice work with this!

File: src/accounts/forms.py
Comment (maxgrove42) [reply src/accounts/forms.py:71]: nice . makes sense!

**Comment by maxgrove42** [comment]
Really great work with this. having less setup and using environment keys makes it much more fluid to deploy. Very impressive. A few small nits on my end but seriously great work

File: src/accounts/views.py
Comment (maxgrove42) [inline src/accounts/views.py:17]: What is the `sender` argument used for here? Things seems to function on my end without it

File: src/accounts/backends.py
Comment (maxgrove42) [inline src/accounts/backends.py:13]: I think you only need the `no-else-return` here right?

File: src/accounts/views.py
Comment (maxgrove42) [inline src/accounts/views.py:32]: 1/ I think you only need the `line-too-long` disabled. And rather than using a disable, you could just split this into two lines at the `'.env'`. 

```python
    env_file = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                            '.env')
```

File: src/accounts/views.py
Comment (maxgrove42) [inline src/accounts/views.py:52]: I think you only need the `no-else-return` disable here

File: src/accounts/views.py
Comment (maxgrove42) [inline src/accounts/views.py:59]: I don't think you need any pylint disable here. There's a trailing space at the end of the line you can get rid of and you can split this into two lines.

```python
            messages.error(request,
                           "Registration Failed!! Password mismatch or too generic") 
```

File: src/accounts/views.py
Comment (maxgrove42) [inline src/accounts/views.py:70]: same here, only need the `no-else-return`

**Review by maxgrove42 — CHANGES_REQUESTED** [review]
few nits re: comments / linter

File: src/accounts/views.py
Comment (maxgrove42) [reply src/accounts/views.py:17]: (it is def possible we need it, but just curious what it is for)

File: src/django_management/settings.py
Comment (TekuriSaiAkhil) [reply src/django_management/settings.py]: Done

File: src/django_management/settings.py
Comment (TekuriSaiAkhil) [reply src/django_management/settings.py]: Done

File: src/accounts/views.py
Comment (TekuriSaiAkhil) [reply src/accounts/views.py]: Updated!! Thanks for the heads up.

File: src/accounts/backends.py
Comment (TekuriSaiAkhil) [reply src/accounts/backends.py]: Updated

File: src/accounts/views.py
Comment (TekuriSaiAkhil) [reply src/accounts/views.py]: Updated

File: src/accounts/views.py
Comment (TekuriSaiAkhil) [reply src/accounts/views.py]: Updated

File: src/accounts/views.py
Comment (TekuriSaiAkhil) [reply src/accounts/views.py]: Updated

File: src/django_management/urls.py
Comment (TekuriSaiAkhil) [reply src/django_management/urls.py:24]: This should be good to close I guess.

File: src/accounts/views.py
Comment (TekuriSaiAkhil) [reply src/accounts/views.py:17]: arguments sender and **kwargs are passed to the signal receiver function to provide flexibility and context about the triggered signal. We don't use these arguments in our function but they are needed to run the code.

**Comment by maxgrove42** [comment]
Approved. Please use `squash and merge` when you get a second review! It's on the drop down of the `Merge Commit` button. Keeps the `develop` branch commit history clean!

Screenshot 2024-10-08 at 5 14 14 PM

**Review by arsalananwar11 — APPROVED** [review]
Looks good to me! Thanks Akhil!

**Pull request merged**
